### ISSUE TITLE
support material theme

### ISSUE THREAD

**Issue by Sispheor** [jimporter/mike#10]
Hi,
When I run `mike install-extras` i get

```
mike: theme 'material' unsupported
```

**Comment by jimporter** [comment]
Yeah, that's intentional. While I suppose I could make an exception for the Material theme, since it's very popular, I didn't want to sign up for maintaining patches for every MkDocs theme out there when creating this project.

`install-extras` isn't strictly required for using mike with a particular theme, though; it just adds the little version selector to pages. Users of a theme can also write their own version selector widget, and theme authors could build it into their themes directly. Finally, third parties (or theme authors) could hook into mike with the `mike.themes` entry point: https://github.com/jimporter/mike/blob/f9fe2cf78dc100f01f3e8caba7c15c5277b01dab/setup.py#L120-L121

**Comment by jimporter** [comment]
Sadly, the author of the Material theme has refused to support Mike since it uses JS for the version selector, citing "progressive enhancement" as Material's primary goal. I disagree fairly strongly with this, since I think "you shouldn't have to touch old docs when building a new version of the docs" is a higher priority. (Injecting the version selector on the server side would be a nice alternative, but that requires the server to be smart enough to do that.)

With that in mind, patches are welcome to support Material for `mike install-extras`. I don't know if I'll have time in the near future to work on it myself, though.

**Comment by adamcharnock** [comment]
I've working on getting mike up and running on the [Lightbus docs](https://lightbus.org/), which are based on mkdocs-material. I set the theme to `readthedocs` then ran `mike install-extras`, before setting it back to `mkdocs-material`. I then updated the `version-select.js` and `version-select.css` files as follows:

(Note that this assumes the docs are served out of the root dir. A little work will be needed to generalise this)

I don't really have the time to turn this into a full on PR right now, but hopefully this can for the basis for one.

## version-select.js

```js
window.addEventListener("DOMContentLoaded", function() {
  function normalizePath(path) {
    var normalized = [];
    path.split("/").forEach(function(bit, i) {
      if (bit === "." || (bit === "" && i !== 0)) {
        return;
      } else if (bit === "..") {
        if (normalized.length === 1 && normalized[0] === "") {
          // We must be trying to .. past the root!
          throw new Error("invalid path");
        } else if (normalized.length === 0 ||
                   normalized[normalized.length - 1] === "..") {
          normalized.push("..");
        } else {
          normalized.pop();
        }
      } else {
        normalized.push(bit);
      }
    });
    return normalized.join("/");
  }

  // `base_url` comes from the base.html template for this theme.
  // Lightbus note: The base_url js variable wasn't immediately obviously
  //                available on the mkdocs-material theme. So we were
  //                simply assume the first part of the URL is the version.
  var CURRENT_VERSION = window.location.pathname.split("/")[1];

  function makeSelect(options, selected) {
    var select = document.createElement("select");

    options.forEach(function(i) {
      var option = new Option(i.text, i.value, undefined,
                              i.value === selected);
      select.add(option);
    });

    return select;
  }

  var xhr = new XMLHttpRequest();
  // Lightbus note: Again, we make assumptions about the path
  xhr.open("GET", "/versions.json");
  xhr.onload = function() {
    var versions = JSON.parse(this.responseText);

    var realVersion = versions.find(function(i) {
      return i.version === CURRENT_VERSION ||
             i.aliases.includes(CURRENT_VERSION);
    }).version;

    var select = makeSelect(versions.map(function(i) {
      return {text: i.title, value: i.version};
    }), realVersion);
    select.addEventListener("change", function(event) {
      window.location.href = "/" + this.value;
    });

    var selectInLi = document.createElement('li');
    selectInLi.appendChild(select);
    selectInLi.className = 'md-nav__item';
    selectInLi.id = 'version-selector';
    var primarySidebarUl = document.querySelector(".md-nav--primary > .md-nav__list");
    var secondarySidebarUl = document.querySelector(".md-nav--primary > .md-nav__list > .md-nav__item--active.md-nav__item--nested .md-nav__list");
    if(secondarySidebarUl) {
      secondarySidebarUl.appendChild(selectInLi);
    } else {
      primarySidebarUl.appendChild(selectInLi);
    }
  };
  xhr.send();
});
```

## version-select.css

```#version-selector {
  padding: .6rem .8rem;
}
.md-nav__item--nested #version-selector {
  padding: .6rem .6rem;
}
#version-selector select {
  display: block;
  width: 100%;
  font-size: .8rem;
}
@media only screen and (min-width:76.1875em) {
.md-nav__item--nested #version-selector {
  padding: .6rem .0rem;
}
  #version-selector {
    padding: .6rem .6rem;
  }
}
```

## Wide

<img width="875" alt="Screenshot 2020-01-17 at 15 06 30" src="https://user-images.githubusercontent.com/138215/72622395-03edef00-393b-11ea-9ecf-5a17b7eb4b0c.png">

## Narrow

<img width="508" alt="Screenshot 2020-01-17 at 15 06 42" src="https://user-images.githubusercontent.com/138215/72622407-0b14fd00-393b-11ea-9c26-0c1b63a1e10d.png">

**Comment by jimporter** [comment]
@adamcharnock Cool, thanks for the patch! I'll try and get this merged in for v0.5, which I'm going to try and release at the same time as MkDocs v1.1 (I had to update the version selector code for the `mkdocs` theme changes).

**Comment by squidfunk** [comment]
Nice work! As you may have noticed, Material 5 is on the horizon and it adds new capabilities of hooking into the theme. Not all things can be supported natively but with some extra leg work (i.e. copy & pasting some CSS and JS) it should be easy to get up and running. For this reason I want to start a new section on the docs called “Recipes” where we could show how to integrate mike with Material besides other similar requests like lightboxes etc.
